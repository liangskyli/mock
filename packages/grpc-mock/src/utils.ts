import type { IPrettierOptions } from '@liangskyli/utils';
import { colors, copyOptions, prettierData } from '@liangskyli/utils';
import fs from 'fs-extra';
import path from 'path';

export const fileTip =
  '// This file is auto generated by @liangskyli/grpc-mock, do not edit!';
export const tslintDisable = '// tslint:disable';
export const longsType = [
  'double',
  'float',
  'int64',
  'uint64',
  'sint64',
  'fixed64',
  'sfixed64',
];
export const PROTO_TYPE_2_TS_TYPE_MAP: Record<string, string> = {
  double: 'number',
  float: 'number',
  int32: 'number',
  int64: 'number',
  uint32: 'number',
  uint64: 'number',
  sint32: 'number',
  sint64: 'number',
  fixed32: 'number',
  fixed64: 'number',
  sfixed32: 'number',
  sfixed64: 'number',
  bool: 'boolean',
  string: 'string',
  bytes: 'string',
};

export const TS_TYPE_2_DEFAULT_MAP: Record<string, any> = {
  number: 0,
  boolean: false,
  string: '""',
  stringNumber: '"1"',
};

export const genSpace = (num: number) => {
  let space = '';
  for (let i = 0; i < num; i++) {
    space += ' ';
  }
  return space;
};

export const firstWordNeedLetter = (str: string): string => {
  const charFirst = str.charCodeAt(0);
  if (
    !(
      (charFirst >= 65 && charFirst <= 90) ||
      (charFirst >= 97 && charFirst <= 122)
    )
  ) {
    //第一个字符不是字母时，补z,保证生成的语法正确
    str = `z${str}`;
  }
  return str;
};

export const firstUpperCaseOfWord = (str: string) => {
  const result: string[] = [];
  str = firstWordNeedLetter(str);
  str
    .replace(/-/gi, '_')
    .split('_')
    .forEach((item) => {
      result.push(
        item.toLowerCase().replace(/^\w|\s\w/g, (w) => w.toUpperCase()),
      );
    });
  return result.join('');
};

export function getImportPath(fromPath: string, toPath: string) {
  let relative = path.relative(path.dirname(fromPath), toPath);
  relative = relative.replace(/\.(js|d\.ts|ts)$/, '').replace(/\\/g, '/');
  if (!/^\./.test(relative)) {
    relative = `./${relative}`;
  }
  return relative;
}

export const packageName = '@liangskyli/grpc-mock';

type IWriteFileOpts = {
  prettierOptions: IPrettierOptions;
  absolutePath: string;
  data: string;
  successTip: string;
};
export const writePrettierFile = (opts: IWriteFileOpts) => {
  const { absolutePath, prettierOptions, data, successTip } = opts;

  fs.writeFileSync(
    absolutePath,
    prettierData(data, copyOptions(prettierOptions)),
  );

  console.info(colors.green(successTip));
};
