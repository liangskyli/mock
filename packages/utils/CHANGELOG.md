# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.0.4](https://github.com/liangskyli/mock/compare/v4.0.3...v4.0.4) (2024-09-30)


### Bug Fixes

* esbuild 0.23.0 to 0.20.1, support node16 ([0eb0972](https://github.com/liangskyli/mock/commit/0eb0972f0d7fd0e4f2f160e7d1669f8a85e4be30))



## [4.0.1](https://github.com/liangskyli/mock/compare/v4.0.0...v4.0.1) (2024-08-03)

**Note:** Version bump only for package @liangskyli/utils





## [4.0.0](https://github.com/liangskyli/mock/compare/v4.0.0-beta.0...v4.0.0) (2024-02-28)

**Note:** Version bump only for package @liangskyli/utils





## [4.0.0-beta.0](https://github.com/liangskyli/mock/compare/v3.2.0-beta.0...v4.0.0-beta.0) (2024-02-26)

**Note:** Version bump only for package @liangskyli/utils





## 3.2.0-beta.0 (2023-12-10)

* feat: http-mock-gen and utils package add "type": "module" as default ([e616e6f](https://github.com/liangskyli/mock/commit/e616e6f))





## 3.1.0 (2023-12-03)

* chore: upgrade dependencies ([f6251cd](https://github.com/liangskyli/mock/commit/f6251cd))
* chore: vitest file rename ([952e3ff](https://github.com/liangskyli/mock/commit/952e3ff))





## [3.0.1](https://github.com/liangskyli/mock/compare/v3.0.0...v3.0.1) (2023-10-21)

**Note:** Version bump only for package @liangskyli/utils





## [3.0.0](https://github.com/liangskyli/mock/compare/v3.0.0-beta.1...v3.0.0) (2023-07-15)

**Note:** Version bump only for package @liangskyli/utils





## [3.0.0-beta.1](https://github.com/liangskyli/mock/compare/v3.0.0-beta.0...v3.0.0-beta.1) (2023-07-15)


### Features

* prettier from dependencies to peerDependencies,support v2 and v3 ([16f48d5](https://github.com/liangskyli/mock/commit/16f48d59c84b654efe9085d5d9e6000aa2a0d49b))



## [3.0.0-beta.0](https://github.com/liangskyli/mock/compare/v2.1.0...v3.0.0-beta.0) (2023-07-12)

**Note:** Version bump only for package @liangskyli/utils





## [2.1.0](https://github.com/liangskyli/mock/compare/v2.0.0...v2.1.0) (2023-06-05)

**Note:** Version bump only for package @liangskyli/utils





## [2.0.0](https://github.com/liangskyli/mock/compare/v2.0.0-beta.5...v2.0.0) (2023-05-13)

**Note:** Version bump only for package @liangskyli/utils





## [2.0.0-beta.5](https://github.com/liangskyli/mock/compare/v2.0.0-beta.4...v2.0.0-beta.5) (2023-05-13)

**Note:** Version bump only for package @liangskyli/utils





## [2.0.0-beta.3](https://github.com/liangskyli/mock/compare/v2.0.0-beta.2...v2.0.0-beta.3) (2023-05-02)


### ⚠ BREAKING CHANGES

* register use Unique key, restore all register

### Features

* register use Unique key, restore all register ([359fe62](https://github.com/liangskyli/mock/commit/359fe62ac5d9d4c0021626757e6c8b9b613020ca))



## [2.0.0-beta.2](https://github.com/liangskyli/mock/compare/v2.0.0-beta.1...v2.0.0-beta.2) (2023-05-01)


### ⚠ BREAKING CHANGES

* prettierData change async to sync

### Code Refactoring

* prettierData change async to sync ([15ff688](https://github.com/liangskyli/mock/commit/15ff688131f8a21ba86be69edea0dc4f88844f2d))



## [2.0.0-beta.1](https://github.com/liangskyli/mock/compare/v2.0.0-beta.0...v2.0.0-beta.1) (2023-04-25)


### ⚠ BREAKING CHANGES

* drop support for Node 14, now support >=16.10.0

### Miscellaneous Chores

* drop support for Node 14 ([8f825b2](https://github.com/liangskyli/mock/commit/8f825b27fed596f7a9e5ab10b2265a75bdcb6504))



## [2.0.0-beta.0](https://github.com/liangskyli/mock/compare/v1.6.0...v2.0.0-beta.0) (2023-04-24)


### Features

* add register, move cross-spawn,ts-node to devDependencies ([ad2c90d](https://github.com/liangskyli/mock/commit/ad2c90db6f288fdff18e63850d46b51e5040a006))
* remove parseRequireDeps function ([8e00fc6](https://github.com/liangskyli/mock/commit/8e00fc6ec25a8a6fa88b1bb8c6926dca092ea440))



## [1.6.0](https://github.com/liangskyli/mock/compare/v1.5.0...v1.6.0) (2023-02-25)


### Features

* prettierOptions type use @liangskyli/utils IPrettierOptions type ([c6a7295](https://github.com/liangskyli/mock/commit/c6a72959bba638d2630a1149fd69405ee47ac731))



## [1.5.0](https://github.com/liangskyli/mock/compare/v1.4.0...v1.5.0) (2022-12-03)

**Note:** Version bump only for package @liangskyli/utils

## [1.4.0](https://github.com/liangskyli/mock/compare/v1.3.0...v1.4.0) (2022-10-04)

**Note:** Version bump only for package @liangskyli/utils

## [1.3.0](https://github.com/liangskyli/mock/compare/v1.2.1...v1.3.0) (2022-08-10)

**Note:** Version bump only for package @liangskyli/utils

## [1.1.3](https://github.com/liangskyli/mock/compare/v1.1.2...v1.1.3) (2022-06-19)

**Note:** Version bump only for package @liangskyli/utils

## [1.1.1](https://github.com/liangskyli/mock/compare/v1.1.0...v1.1.1) (2022-06-02)

**Note:** Version bump only for package @liangskyli/utils

## [1.1.0](https://github.com/liangskyli/mock/compare/v1.0.0...v1.1.0) (2022-05-28)

### Features

- console colors show ([614a473](https://github.com/liangskyli/mock/commit/614a4738612f471598263e562d14fce5c215c17b))
- optimization ts type and code ([d48f9ce](https://github.com/liangskyli/mock/commit/d48f9ce9c9921a961fc8bcba54c328643cdb5abb))

## [1.0.0](https://github.com/liangskyli/mock/compare/v1.0.0-beta.0...v1.0.0) (2022-05-15)

**Note:** Version bump only for package @liangskyli/utils

## [1.0.0-beta.0](https://github.com/liangskyli/mock/compare/v0.12.0...v1.0.0-beta.0) (2022-05-15)

### Features

- **utils:** add getRelativePath ([68e9463](https://github.com/liangskyli/mock/commit/68e946334995f3cf0da11abb94049bed1570689d))

## [0.12.0](https://github.com/liangskyli/mock/compare/v0.11.0...v0.12.0) (2022-04-24)

**Note:** Version bump only for package @liangskyli/utils

## [0.11.0](https://github.com/liangskyli/mock/compare/v0.10.0-beta.1...v0.11.0) (2022-04-17)

**Note:** Version bump only for package @liangskyli/utils

## [0.10.0-beta.0](https://github.com/liangskyli/mock/compare/v0.9.1-beta.0...v0.10.0-beta.0) (2022-04-04)

### Features

- http-mock-gen support separation of custom data and generated data ([585e59d](https://github.com/liangskyli/mock/commit/585e59db3ccf208203087751e1db27b627d6d0f9))

### [0.9.1-beta.0](https://github.com/liangskyli/mock/compare/v0.9.0...v0.9.1-beta.0) (2022-02-25)

**Note:** Version bump only for package @liangskyli/utils

## [0.9.0](https://github.com/liangskyli/mock/compare/v0.9.0-beta.5...v0.9.0) (2022-01-27)

**Note:** Version bump only for package @liangskyli/utils

## [0.9.0-beta.0](https://github.com/liangskyli/mock/compare/v0.8.1...v0.9.0-beta.0) (2022-01-27)

**Note:** Version bump only for package @liangskyli/utils

## [0.8.0](https://github.com/liangskyli/mock/compare/v0.8.0-beta.0...v0.8.0) (2022-01-14)

### Bug Fixes

- window path ([50014b7](https://github.com/liangskyli/mock/commit/50014b7dd64b1a9405ce23e7ae32987206b18764))

## [0.8.0-beta.0](https://github.com/liangskyli/mock/compare/v0.7.0...v0.8.0-beta.0) (2022-01-13)

**Note:** Version bump only for package @liangskyli/utils
