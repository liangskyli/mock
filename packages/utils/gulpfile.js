const gulp = require('gulp');
const del = require('del');
const through = require('through2');

function clean() {
  return del('./lib/**');
}

function copyMetaFiles() {
  return gulp.src(['./README.md']).pipe(gulp.dest('./lib/'));
}

function generateCliPackageJSON() {
  return gulp
    .src('./package.json')
    .pipe(
      through.obj((file, enc, cb) => {
        const rawJSON = file.contents.toString();
        const parsed = JSON.parse(rawJSON);
        parsed.main = parsed.main.replace('./lib/', './');
        parsed.module = parsed.module.replace('./lib/', './');
        parsed.types = parsed.types.replace('./lib/', './');
        parsed.typings = parsed.typings.replace('./lib/', './');
        delete parsed.scripts;
        delete parsed.devDependencies;
        delete parsed.publishConfig;
        delete parsed.files;
        const stringified = JSON.stringify(parsed, null, 2);
        file.contents = Buffer.from(stringified);
        cb(null, file);
      }),
    )
    .pipe(gulp.dest('lib/'));
}

exports.default = gulp.series(clean, copyMetaFiles, generateCliPackageJSON);
