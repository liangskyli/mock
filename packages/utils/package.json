{
  "name": "@liangskyli/utils",
  "version": "4.0.1",
  "description": "@liangskyli/utils",
  "keywords": [],
  "homepage": "https://github.com/liangskyli/mock#readme",
  "bugs": {
    "url": "https://github.com/liangskyli/mock/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/liangskyli/mock.git"
  },
  "license": "MIT",
  "author": "liangsky <640634387@qq.com>",
  "type": "module",
  "main": "./lib/index.cjs",
  "module": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "typings": "./lib/index.d.ts",
  "exports": {
    ".": {
      "types": "./lib/index.d.ts",
      "import": "./lib/index.js",
      "require": "./lib/index.cjs"
    }
  },
  "files": [
    "lib",
    "CHANGELOG.md",
    "LICENSE",
    "README.md",
    "package.json"
  ],
  "scripts": {
    "test": "vitest run",
    "coverage": "vitest run --coverage",
    "build": "rollup --config=./rollup.config.js",
    "update:deps": "pnpm update --interactive --latest"
  },
  "dependencies": {
    "@types/lodash": "^4.17.7",
    "address": "^2.0.3",
    "colors": "^1.4.0",
    "debug": "^4.3.6",
    "esbuild": "^0.20.1",
    "esbuild-register": "^3.6.0",
    "fs-extra": "^11.2.0",
    "lodash": "^4.17.21",
    "signale": "^1.4.0"
  },
  "devDependencies": {
    "@types/debug": "^4.1.12",
    "@types/fs-extra": "^11.0.4",
    "@types/signale": "^1.4.7"
  },
  "peerDependencies": {
    "prettier": "^2.0.0 || ^3.0.0"
  },
  "engines": {
    "node": "^14.18.0 || >=16.0.0"
  }
}
