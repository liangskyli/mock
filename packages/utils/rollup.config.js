import path from 'path';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import typescript from 'rollup-plugin-typescript2';
import json from '@rollup/plugin-json';
import packageJSON from './package.json';

const extensions = ['.js', '.ts', '.tsx'];
const getPath = (_path) => path.resolve(__dirname, _path);

// ts
const tsPlugin = typescript({
  tsconfig: getPath('./tsconfig.json'), // 导入本地ts配置
});

// 基础配置
const commonConf = {
  input: getPath('./src/index.ts'),
  external: [
    ...Object.keys(packageJSON.dependencies || {}),
    ...Object.keys(packageJSON.peerDependencies || {}),
  ],
  plugins: [nodeResolve({ extensions }), tsPlugin, commonjs(), json()],
};
// 需要导出的模块类型
const outputMap = [
  {
    file: packageJSON.main, // 通用模块
    format: 'cjs',
  },
  {
    file: packageJSON.module, // es6模块
    format: 'es',
  },
];

const buildConf = (options) => Object.assign({}, commonConf, options);

export default outputMap.map((output) =>
  buildConf({ output: { name: packageJSON.name, ...output } }),
);
