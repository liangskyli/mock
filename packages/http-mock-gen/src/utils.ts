import type { IGenTsDataOpts } from '@liangskyli/openapi-gen-ts';
import { colors, copyOptions, lodash, prettierData } from '@liangskyli/utils';
import fs from 'fs-extra';

export const fileTip =
  '// This file is auto generated by @liangskyli/http-mock-gen, do not edit!';

export const packageName = '@liangskyli/http-mock-gen';

type AssignCustomizer = (
  objectValue: any,
  sourceValue: any,
  key?: string,
  object?: any,
  source?: any,
) => any;

const customizer: AssignCustomizer = (value, srcValue) => {
  if (lodash.isObject(value)) {
    if (lodash.isArray(value)) {
      if (lodash.isArray(srcValue) && srcValue.length > 0) {
        return value.map((item) => {
          if (lodash.isObject(item)) {
            return lodash.mergeWith(item, srcValue[0], customizer);
          } else {
            // simple array data, no merge, use new value
            return item;
          }
        });
      } else {
        return value;
      }
    } else {
      return lodash.mergeWith(value, srcValue, customizer);
    }
  } else {
    return value === undefined ? srcValue : value;
  }
};

const mergeUndefined = (temp: any, custom: any) => {
  if (lodash.isObject(custom)) {
    if (lodash.isArray(custom)) {
      custom.forEach((item, index) => {
        if (lodash.isObject(item)) {
          mergeUndefined(temp[index], item);
        } else {
          if (item === undefined) {
            temp[index] = undefined;
          }
        }
      });
    } else {
      Object.keys(custom).forEach((key) => {
        const item = (custom as any)[key];
        if (lodash.isObject(item)) {
          mergeUndefined(temp[key], item);
        } else {
          if (item === undefined) {
            temp[key] = undefined;
          }
        }
      });
    }
  }
  return temp;
};

// 自定义数据和默认数据合并
export const mergeObject = (object: any, source: any) => {
  const tempData = lodash.mergeWith(
    lodash.cloneDeep(object),
    source,
    customizer,
  );
  // 自定义undefined数据覆盖默认数据
  const result = mergeUndefined(tempData, object);
  return result;
};

export const methodList = [
  'get',
  'put',
  'post',
  'delete',
  'options',
  'head',
  'patch',
  'trace',
];

type IWriteFileOpts = {
  prettierOptions: IGenTsDataOpts['prettierOptions'];
  absolutePath: string;
  data: string;
  successTip: string;
};
export const writePrettierFile = (opts: IWriteFileOpts) => {
  const { absolutePath, prettierOptions, data, successTip } = opts;

  fs.writeFileSync(
    absolutePath,
    prettierData(data, copyOptions(prettierOptions)),
  );

  console.info(colors.green(successTip));
};

export const getMethodData = (itemValue: any) => {
  // support all openapi method
  let method = '';
  let data: any = '';
  let responseMediaType = '';
  methodList.forEach((item) => {
    if (itemValue?.[item]) {
      method = item;
      const responsesProperties = itemValue[item]?.responses;
      // first use responses 200, then use responses default
      let responseCode: '200' | 'default' = '200';
      if (!responsesProperties?.['200']) {
        responseCode = 'default';
      }
      // first use responses 200, then use responses default
      const responsesContent = responsesProperties?.[responseCode]?.content;
      if (responsesContent) {
        // responses content only use first key
        responseMediaType = Object.keys(responsesContent)[0];
      }

      data = responsesContent?.[responseMediaType] ?? {};
    }
  });
  return { method, data };
};
