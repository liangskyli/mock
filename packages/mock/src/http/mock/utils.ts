import { createDebug, winPath } from '@liangskyli/utils';
import assert from 'assert';
import bodyParser from 'body-parser';
import type { NextFunction, Request, RequestHandler } from 'express';
import { globSync } from 'glob';
import multer from 'multer';
import { existsSync } from 'node:fs';
import { createRequire } from 'node:module';
import { join } from 'node:path';
import { pathToRegexp } from 'path-to-regexp';

const require = createRequire(import.meta.url);

// support all openapi method
const VALID_METHODS = [
  'get',
  'put',
  'post',
  'delete',
  'options',
  'head',
  'patch',
  'trace',
];
const BODY_PARSED_METHODS = VALID_METHODS.filter((item) => item !== 'get');

const debug = createDebug('mock:utils');

interface IGetMockPaths {
  cwd: string;
  ignore?: string[];
}

export interface IMockDataItem {
  method: string;
  path: string;
  pathToRegexpData: ReturnType<typeof pathToRegexp>;
  handler: RequestHandler;
}

export interface IGetMockDataResult {
  mockData: IMockDataItem[];
  mockPaths: string[];
  mockWatcherPaths: string[];
}

export function getMockConfig(files: string[]) {
  return files.reduce<Record<string, any>>((memo, mockFile) => {
    try {
      const m = require(mockFile);
      memo = {
        ...memo,
        ...(m.default || m),
      };
      return memo;
    } catch (e: any) {
      throw new Error(e.stack);
    }
  }, {});
}

export const cleanRequireCache = (paths: string[]) => {
  Object.keys(require.cache).forEach((file) => {
    if (
      paths.some((path) => {
        return winPath(file).indexOf(path) > -1;
      })
    ) {
      delete require.cache[file];
    }
  });
};

function parseKey(key: string) {
  let method = 'get';
  let path = key;
  if (/\s+/.test(key)) {
    const splited = key.split(/\s+/);
    method = splited[0].toLowerCase();
    path = splited[1];
  }
  assert(
    VALID_METHODS.includes(method),
    `Invalid method ${method} for path ${path}, please check your mock files.`,
  );
  return {
    method,
    path,
  };
}

function createHandler(method: any, path: any, handler: any) {
  return function (req: Request, res: any, next: NextFunction) {
    function sendData() {
      if (typeof handler === 'function') {
        multer().any()(req, res, () => {
          handler(req, res, next);
        });
      } else {
        res.json(handler);
      }
    }

    if (BODY_PARSED_METHODS.includes(method)) {
      bodyParser.json({ limit: '5mb', strict: false })(req, res, () => {
        bodyParser.urlencoded({ limit: '5mb', extended: true })(
          req,
          res as any,
          () => {
            sendData();
          },
        );
      });
    } else {
      sendData();
    }
  } as unknown as RequestHandler;
}

function normalizeConfig(config: Record<string, any>) {
  return Object.keys(config).reduce<IMockDataItem[]>((memo, key) => {
    const handler = config[key];
    const type = typeof handler;
    assert(
      type === 'function' || type === 'object',
      `mock value of ${key} should be function or object, but got ${type}`,
    );
    const { method, path } = parseKey(key);
    const pathToRegexpData = pathToRegexp(path);
    memo.push({
      method,
      path,
      pathToRegexpData,
      handler: createHandler(method, path, handler),
    });
    return memo;
  }, []);
}

function decodeParam(val: any) {
  if (typeof val !== 'string' || val.length === 0) {
    return val;
  }
  try {
    return decodeURIComponent(val);
  } catch (err) {
    if (err instanceof URIError) {
      err.message = `Failed to decode param ' ${val} '`;
      // @ts-expect-error have property
      err.status = 400;
      // @ts-expect-error have property
      err.statusCode = 400;
    }
    throw err;
  }
}

/**
 * mock/*
 * .umirc.mock.js
 * .umirc.mock.ts
 * src/** or pages/**
 *
 * @param opts
 */
export const getMockData: (opts: IGetMockPaths) => IGetMockDataResult = (
  opts,
) => {
  const { cwd, ignore = [] } = opts;
  const mockPaths = [
    ...(globSync('mock/**/*.[jt]s', {
      cwd,
      ignore,
    }) || []),
    ...(globSync('**/_mock.[jt]s', {
      cwd,
      ignore,
    }) || []),
    '.umirc.mock.js',
    '.umirc.mock.ts',
  ]
    .map((path) => join(cwd, path))
    .filter((path) => path && existsSync(path))
    .map((path) => winPath(path));

  debug(`load mock data including files ${JSON.stringify(mockPaths)}`);

  // get mock data
  const mockData = normalizeConfig(getMockConfig(mockPaths));

  const mockWatcherPaths = [...(mockPaths || []), join(cwd, 'mock')]
    .filter((path) => path && existsSync(path))
    .map((path) => winPath(path));

  return {
    mockData,
    mockPaths,
    mockWatcherPaths,
  };
};

export const matchMock = (
  req: Request,
  mockData: IMockDataItem[],
): IMockDataItem | undefined => {
  const { path: targetPath, method } = req;
  const targetMethod = method.toLowerCase();

  for (const mock of mockData) {
    const {
      method: mockMethod,
      pathToRegexpData: { regexp, keys },
    } = mock;
    if (mockMethod === targetMethod) {
      const match = regexp.exec(targetPath);
      if (match) {
        const params: Record<any, any> = {};
        for (let i = 1; i < match.length; i += 1) {
          const key = keys[i - 1];
          const prop = key.name;
          const val = decodeParam(match[i]);
          if (
            val !== undefined ||
            !Object.prototype.hasOwnProperty.call(params, prop)
          ) {
            params[prop] = val;
          }
        }
        req.params = params;
        return mock;
      }
    }
  }
  return undefined;
};
