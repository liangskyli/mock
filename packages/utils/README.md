# `@liangskyli/utils`

> 工具库

## Usage

```
import { getAbsolutePath, prettierData, copyOptions } from '@liangskyli/utils';

```
