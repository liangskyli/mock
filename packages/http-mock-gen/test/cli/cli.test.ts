import { winPath } from '@liangskyli/utils';
import { execa } from 'execa';
import fs from 'fs-extra';
import { URL } from 'node:url';
import path from 'path';
import { describe, expect, test } from 'vitest';

const cwd = new URL('../../', import.meta.url);
const cmd = 'node';
describe('CLI', () => {
  test('CLI test', async () => {
    await fs.remove(
      path.join(__dirname, '../all-gen-dirs/gen-mock-cli/mock/custom-data'),
    );
    const { stdout } = await execa(
      cmd,
      ['./bin/index.js', '-c', './test/cli/mock.config.cli.ts'],
      {
        cwd,
      },
    );
    await expect(winPath(stdout)).toMatchSnapshot();
  });
});
