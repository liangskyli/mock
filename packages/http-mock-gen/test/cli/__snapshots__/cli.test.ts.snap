// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI > CLI test 1`] = `
"[32muse configFile path: ./test/cli/mock.config.cli.ts[39m
[32mClean mock dir: test/all-gen-dirs/gen-mock-cli/mock[39m
[32mClean schema-api dir: test/all-gen-dirs/gen-mock-cli/mock/schema-api[39m
[32mGenerate schema-api/ts-schema.ts success[39m
[32mGenerate schema-api/schema.json success[39m
[32mGenerate schema-api/request.ts file success[39m
[32mGenerate schema-api/interface-api.ts file success[39m
[32mGenerate schema-api/request-api.ts file success[39m
[32mGenerate mock/mock-data.json success[39m
[32mGenerate mock/custom-data/index.ts file success[39m
[32mGenerate mock/custom-data/template-data.ts file success[39m
[32mGenerate mock/interface-mock-data.ts file success[39m"
`;
